### PR TITLE
Add cropping to layout

### PR DESCRIPTION
**Pull request by WojciechBarczynski**
`crop` example was added for crop testing. It's meant to be removed in the final version. Cropping values are hard-coded in `compositor_render/src/transformations/layout.rs`, because cropping is currently not part of the user-facing API and implementing it in the layout engine is outside of the scope of this PR.

Tests and lint failures are expected, due to hardcoded values / `todo!()` macros.

Known issues:
- A silly shader breaks this cropping. I'll fix it in the follow-up PR

### CODE REVIEW

File: compositor_render/src/transformations/layout/apply_layouts.wgsl [hunk 1]
```diff
@@ -12,9 +12,10 @@ struct VertexOutput {
 
 
 struct Layout {
-    position_transformation: mat4x4<f32>,
-    texture_id: i32,
+    vertices_transformation: mat4x4<f32>,
+    texture_coord_transformation: mat4x4<f32>,
```
Comment (WojciechBarczynski) [inline compositor_render/src/transformations/layout/apply_layouts.wgsl:16]: Technically we need only `mat3x3<f32>`, but the memory layout of `mat3x3` requires adding padding to 16 bytes after each column, so I used `mat4x4` for convenience.

File: examples/crop.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
+use anyhow::Result;
```
Comment (wkozyra95) [inline examples/crop.rs:1]: delete this

File: compositor_render/src/transformations/layout.rs [hunk 3]
```diff
@@ -118,22 +139,71 @@ impl LayoutNode {
...
+                        -1,
+                        RGBAColor(0, 0, 0, 0),
+                        Resolution {
+                            width: 0,
+                            height: 0,
+                        },
+                    ),
                 };
+
+                // TODO remove this - it's hard coded for crop example
```
Comment (wkozyra95) [inline compositor_render/src/transformations/layout.rs:183]: delete this

File: compositor_render/src/transformations/layout.rs [hunk 4]
```diff
@@ -57,8 +65,17 @@ pub struct NestedLayout {
 impl NestedLayout {
     pub fn flatten(mut self, child_index_offset: usize) -> Vec<Layout> {
         let mut child_index_offset = child_index_offset;
-        if let LayoutContent::ChildNode(index) = self.layout.content {
-            self.layout.content = LayoutContent::ChildNode(index + child_index_offset);
+        if let LayoutContent::ChildNode { index, .. } = self.layout.content {
+            // TODO fix this
```
Comment (wkozyra95) [inline compositor_render/src/transformations/layout.rs:69]: should be fixed after rebase

File: compositor_render/src/transformations/layout/transformation_matrices.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
+use compositor_common::scene::Resolution;
+use nalgebra_glm::{rotate_z, scale, translate, vec3, Mat4, Vec3};
+
+use super::Layout;
+
+impl Layout {
```
Comment (wkozyra95) [inline compositor_render/src/transformations/layout/transformation_matrices.rs:6]: Note that this will be RenderLayout after rebase

**Review by wkozyra95 — COMMENTED** [review]
LGTM, most issues should be resolved after the rebase

**Review by noituri — APPROVED** [review]

**Review by wkozyra95 — APPROVED** [review]
It's fine if you want to leave it , but I would prefer if you did not change a submodule commit if you are not changing anything in tests.

**Comment by WojciechBarczynski** [comment]
> It's fine if you want to leave it, but I would prefer if you did not change a submodule commit if you are not changing anything in tests.

I just used the last main ref of the snapshots repository, but I'll revert it.

**Pull request merged**